File: src/tools/logit/logit.cc
//
// This file is part of Gambit
// Copyright (c) 1994-2025, The Gambit Project (https://www.gambit-project.org)
//
// FILE: src/tools/logit/logit.cc
// Command-line driver program for quantal response equilibrium tracing and
// maximum likelihood estimation.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <getopt.h>
#include "gambit.h"
#include "solvers/logit/logit.h"

using namespace Gambit;

void PrintBanner(std::ostream &p_stream)
{
  p_stream << "Compute a branch of the logit equilibrium correspondence\n";
  p_stream << "Gambit version " VERSION ", ";
  p_stream << "Copyright (C) 1994-2025, The Gambit Project\n";
  p_stream << "This is free software, distributed under the GNU GPL\n\n";
}

void PrintHelp(char *progname)
{
  PrintBanner(std::cerr);
  std::cerr << "Usage: " << progname << " [OPTIONS] [file]\n";
  std::cerr << "If file is not specified, attempts to read game from standard input.\n";

  std::cerr << "Options:\n";
  std::cerr << "  -d DECIMALS      show equilibria as floating point with DECIMALS digits\n";
  std::cerr << "  -s STEP          initial stepsize (default is .03)\n";
  std::cerr << "  -a ACCEL         maximum acceleration (default is 1.1)\n";
  std::cerr << "  -m MAXREGRET     maximum regret acceptable as a proportion of range of\n";
  std::cerr << "                   payoffs in the game\n";
  std::cerr << "  -l LAMBDA        compute QRE at `lambda` accurately\n";
  std::cerr << "  -L FILE          compute maximum likelihood estimates;\n";
  std::cerr << "                   read strategy frequencies from FILE\n";
  std::cerr << "  -h, --help       print this help message\n";
  std::cerr << "  -q               quiet mode (suppresses banner)\n";
  std::cerr << "  -e               print only the terminal equilibrium\n";
  std::cerr << "                   (default is to print the entire branch)\n";
  std::cerr << "  -v, --version    print version information\n";
  exit(1);
}

//
// Read in a comma-separated values list of observed data values
//
bool ReadProfile(std::istream &p_stream, MixedStrategyProfile<double> &p_profile)
{
  for (size_t i = 1; i <= p_profile.MixedProfileLength(); i++) {
    if (p_stream.eof() || p_stream.bad()) {
      return false;
    }

    p_stream >> p_profile[i];
    if (i < p_profile.MixedProfileLength()) {
      char comma;
      p_stream >> comma;
    }
  }
  // Read in the rest of the line and discard
  std::string foo;
  std::getline(p_stream, foo);
  return true;
}

template <class T>
void PrintProfile(std::ostream &p_stream, int p_decimals, const T &p_profile, bool p_nash = false)
{
  if (p_nash) {
    p_stream << "NE";
  }
  else {
    p_stream.setf(std::ios::fixed);
    p_stream << std::setprecision(p_decimals) << p_profile.GetLambda();
    p_stream.unsetf(std::ios::fixed);
  }
  for (size_t i = 1; i <= p_profile.size(); i++) {
    p_stream << "," << std::setprecision(p_decimals) << p_profile[i];
  }
  if (p_profile.GetLogLike() <= 0.0) {
    p_stream.setf(std::ios::fixed);
    p_stream << "," << std::setprecision(p_decimals) << p_profile.GetLogLike();
    p_stream.unsetf(std::ios::fixed);
  }
  p_stream << std::endl;
}

int main(int argc, char *argv[])
{
  opterr = 0;

  bool quiet = false, useStrategic = false;
  const double maxLambda = 1000000.0;
  double maxregret = 1.0e-8;
  std::string mleFile;
  double maxDecel = 1.1;
  double hStart = 0.03;
  std::list<double> targetLambda;
  bool fullGraph = true;
  int decimals = 6;

  int long_opt_index = 0;
  struct option long_options[] = {
      {"help", 0, nullptr, 'h'}, {"version", 0, nullptr, 'v'}, {nullptr, 0, nullptr, 0}};
  int c;
  while ((c = getopt_long(argc, argv, "d:s:a:m:vqehSL:p:l:", long_options, &long_opt_index)) !=
         -1) {
    switch (c) {
    case 'v':
      PrintBanner(std::cerr);
      exit(1);
    case 'q':
      quiet = true;
      break;
    case 'd':
      decimals = atoi(optarg);
      break;
    case 's':
      hStart = atof(optarg);
      break;
    case 'a':
      maxDecel = atof(optarg);
      break;
    case 'm':
      maxregret = atof(optarg);
      break;
    case 'e':
      fullGraph = false;
      break;
    case 'h':
      PrintHelp(argv[0]);
      break;
    case 'S':
      useStrategic = true;
      break;
    case 'L':
      mleFile = optarg;
      break;
    case 'l':
      targetLambda.push_back(atof(optarg));
      break;
    case '?':
      if (isprint(optopt)) {
        std::cerr << argv[0] << ": Unknown option `-" << ((char)optopt) << "'.\n";
      }
      else {
        std::cerr << argv[0] << ": Unknown option character `\\x" << optopt << "`.\n";
      }
      return 1;
    default:
      abort();
    }
  }

  if (!quiet) {
    PrintBanner(std::cerr);
  }

  std::istream *input_stream = &std::cin;
  std::ifstream file_stream;
  if (optind < argc) {
    file_stream.open(argv[optind]);
    if (!file_stream.is_open()) {
      std::ostringstream error_message;
      error_message << argv[0] << ": " << argv[optind];
      perror(error_message.str().c_str());
      exit(1);
    }
    input_stream = &file_stream;
  }

  try {
    const Game game = ReadGame(*input_stream);
    if (!game->IsPerfectRecall()) {
      throw UndefinedException(
          "Computing equilibria of games with imperfect recall is not supported.");
    }

    if (!mleFile.empty() && (!game->IsTree() || useStrategic)) {
      MixedStrategyProfile<double> frequencies(game->NewMixedStrategyProfile(0.0));
      std::ifstream mleData(mleFile.c_str());
      ReadProfile(mleData, frequencies);

      auto printer = [fullGraph, decimals](const LogitQREMixedStrategyProfile &p) {
        if (fullGraph) {
          PrintProfile(std::cout, decimals, p);
        }
      };
      auto result =
          LogitStrategyEstimate(frequencies, maxLambda, 1.0, false, hStart, maxDecel, printer);
      PrintProfile(std::cout, decimals, result);
      return 0;
    }

    if (!game->IsTree() || useStrategic) {
      auto printer = [fullGraph, decimals](const LogitQREMixedStrategyProfile &p) {
        if (fullGraph) {
          PrintProfile(std::cout, decimals, p);
        }
      };
      const LogitQREMixedStrategyProfile start(game);
      if (!targetLambda.empty()) {
        auto result =
            LogitStrategySolveLambda(start, targetLambda, 1.0, hStart, maxDecel, printer);
        for (auto &profile : result) {
          PrintProfile(std::cout, decimals, profile);
        }
      }
      else {
        auto result = LogitStrategySolve(start, maxregret, 1.0, hStart, maxDecel, printer);
        PrintProfile(std::cout, decimals, result.back(), true);
      }
    }
    else {
      auto printer = [fullGraph, decimals](const LogitQREMixedBehaviorProfile &p) {
        if (fullGraph) {
          PrintProfile(std::cout, decimals, p);
        }
      };
      const LogitQREMixedBehaviorProfile start(game);
      if (!targetLambda.empty()) {
        auto result =
            LogitBehaviorSolveLambda(start, targetLambda, 1.0, hStart, maxDecel, printer);
        for (auto &profile : result) {
          PrintProfile(std::cout, decimals, profile);
        }
      }
      else {
        auto result = LogitBehaviorSolve(start, maxregret, 1.0, hStart, maxDecel, printer);
        PrintProfile(std::cout, decimals, result.back(), true);
      }
    }
    return 0;
  }
  catch (std::runtime_error &e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: src/pygambit/cli/logit.py
#
# This file is part of Gambit
# Copyright (c) 1994-2026, The Gambit Project (https://www.gambit-project.org)
#
# FILE: src/pygambit/cli/logit.py
# Command-line driver program for quantal response equilibrium tracing
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
"""Command-line driver program for quantal response equilibrium tracing."""

from __future__ import annotations

import pathlib

import click

import pygambit as gbt

from .common import (
    handle_errors,
    open_game_file,
    print_banner,
    read_game,
    render_profile_csv,
    version_option,
)

DESCRIPTION = "Compute a branch of the logit equilibrium correspondence"
PROG_NAME = "gambit-logit"


def _render_qre_row(qre, decimals: int) -> str:
    """Render a QRE point (anything exposing `lam`, `profile`, and `log_like`, as
    `LogitQREMixedStrategyProfile`/`LogitQREMixedBehaviorProfile` and
    `LogitQREMixedStrategyFitResult`/`LogitQREMixedBehaviorFitResult` all do) as a
    single comma-separated line, matching `gambit-logit`'s own `PrintProfile`: the
    lambda label is fixed-point, the profile values use `decimals` significant
    digits, and the log-likelihood is appended, fixed-point, when it is meaningful
    (non-positive).
    """
    label = f"{qre.lam:.{decimals}f}"
    text = render_profile_csv(qre.profile, label, decimals, fixed=False)
    if qre.log_like <= 0.0:
        text += f",{qre.log_like:.{decimals}f}"
    return text


def _read_frequencies(path: str, game: gbt.Game) -> gbt.MixedStrategyProfileDouble:
    """Read observed strategy frequencies for maximum-likelihood estimation: a flat,
    comma-separated list of counts, one per strategy, in the same order as a profile's
    CSV row, matching the C++ tool's `ReadProfile`.
    """
    count = sum(len(list(player.strategies)) for player in game.players)
    try:
        fields = pathlib.Path(path).read_text().split(",")
        values = [float(fields[i]) for i in range(count)]
    except (OSError, ValueError, IndexError):
        raise ValueError(f"Error reading strategy frequencies from '{path}'.") from None
    frequencies = game.mixed_strategy_profile(rational=False)
    it = iter(values)
    for player in game.players:
        frequencies[player.label] = {s: next(it) for s in player.strategies}
    return frequencies


@click.command(
    context_settings={"help_option_names": ["-h", "--help"]},
    help=(
        f"{DESCRIPTION}.\n\n"
        "Reads a game from FILE, or from standard input if FILE is not specified."
    ),
)
@click.argument("file", required=False, default=None)
@click.option(
    "-d",
    "decimals",
    default=6,
    show_default=True,
    type=int,
    help="show equilibria as floating point with DECIMALS digits",
)
@click.option("-S", "--strategic", is_flag=True, help="use strategic game")
@click.option(
    "-s",
    "first_step",
    default=0.03,
    show_default=True,
    type=float,
    help="initial stepsize for the tracing procedure",
)
@click.option(
    "-a",
    "max_accel",
    default=1.1,
    show_default=True,
    type=float,
    help="maximum acceleration of the step size during the tracing procedure",
)
@click.option(
    "-m",
    "maxregret",
    default=1.0e-8,
    show_default=True,
    type=float,
    help="maximum regret acceptable as a proportion of the range of payoffs in the game",
)
@click.option(
    "-l",
    "target_lambda",
    multiple=True,
    type=float,
    help="compute the QRE at LAMBDA accurately; may be specified multiple times",
)
@click.option(
    "-L",
    "mle_file",
    type=str,
    default=None,
    help="compute maximum likelihood estimate; read strategy frequencies from FILE",
)
@click.option(
    "-e",
    "--terminal-only",
    is_flag=True,
    help="print only the terminal equilibrium (default is to print the entire branch)",
)
@click.option("-q", "--quiet", is_flag=True, help="quiet mode (suppresses banner)")
@version_option(DESCRIPTION)
@handle_errors
def main(
    file: str | None,
    decimals: int,
    strategic: bool,
    first_step: float,
    max_accel: float,
    maxregret: float,
    target_lambda: tuple[float, ...],
    mle_file: str | None,
    terminal_only: bool,
    quiet: bool,
) -> None:
    if not quiet:
        print_banner(DESCRIPTION)
    game = read_game(open_game_file(file, PROG_NAME))
    if not game.is_perfect_recall:
        raise ValueError("Computing equilibria of games with imperfect recall is not supported.")

    def stream(qre) -> None:
        click.echo(_render_qre_row(qre, decimals))

    event_callback = None if terminal_only else stream

    # Maximum-likelihood estimation, like the C++ tool, is only defined over the
    # strategic representation, since the observed frequencies are read as a flat
    # list of strategy counts.
    if mle_file is not None and (strategic or not game.is_tree):
        frequencies = _read_frequencies(mle_file, game)
        result = gbt.qre.logit_estimate(
            frequencies,
            first_step=first_step,
            max_accel=max_accel,
            event_callback=event_callback,
        )
        click.echo(_render_qre_row(result, decimals))
        return

    if target_lambda:
        for qre in gbt.qre.logit_solve_lambda(
            game,
            list(target_lambda),
            use_strategic=strategic,
            first_step=first_step,
            max_accel=max_accel,
            event_callback=event_callback,
        ):
            click.echo(_render_qre_row(qre, decimals))
        return

    if terminal_only:
        result = gbt.nash.logit_solve(
            game,
            use_strategic=strategic,
            maxregret=maxregret,
            first_step=first_step,
            max_accel=max_accel,
        )
        click.echo(render_profile_csv(result.equilibria[-1], "NE", decimals, fixed=False))
        return

    result = gbt.nash.logit_solve(
        game,
        use_strategic=strategic,
        maxregret=maxregret,
        first_step=first_step,
        max_accel=max_accel,
        event_callback=stream,
    )
    click.echo(render_profile_csv(result.equilibria[-1], "NE", decimals, fixed=False))


if __name__ == "__main__":
    main()
